crypto/crypto/src/x25519.rs

```diff
@@ -1,340 +1,151 @@
-//! let (private_key2, public_key2) = X25519StaticPrivateKey::generate_keypair_hybrid(Some(salt), &seed, Some(info));
-//! assert_ne!(public_key1, public_key2);
+//! // Deserialize an hexadecimal private or public key
+//! use libra_crypto::traits::ValidKeyStringExt;
+//! # fn main() -> Result<(), libra_crypto::traits::CryptoMaterialError> {
+//! let private_key = "404acc8ec6a0f18df7359a6ee7823f19dd95616b10fed8bdb0de030e891b945a";
+//! let private_key = x25519::PrivateKeyBytes::from_encoded_string(&private_key)?;
+//! let public_key = "080e287879c918794170e258bfaddd75acac5b3e350419044655e4983a487120";
+//! let public_key = x25519::PublicKeyBytes::from_encoded_string(&public_key)?;
+//! # Ok(())
+//! # }
 //! ```
+//!
 
-use crate::{hkdf::Hkdf, traits::*};
-use libra_crypto_derive::{Deref, DeserializeKey, SerializeKey, SilentDebug, SilentDisplay};
-use rand::{rngs::EntropyRng, RngCore};
-use sha2::Sha256;
-use std::{convert::TryFrom, ops::Deref};
-
-/// TODO: move traits to the right file (possibly traits.rs)
-
-/// Key interfaces for Diffie-Hellman key exchange protocol build on top
-/// of the key APIs in traits.rs
-
-/// x25519 implementation
-
-/// The length of the DHPublicKey
-pub const X25519_PUBLIC_KEY_LENGTH: usize = 32;
-/// The length of the DHPrivateKey
-pub const X25519_PRIVATE_KEY_LENGTH: usize = 32;
-
-/// An x25519 ephemeral private (secret) key
-#[derive(SilentDisplay, SilentDebug)]
-pub struct X25519EphemeralPrivateKey(x25519_dalek::EphemeralSecret);
-
-/// An x25519 static private (secret) key
-#[derive(Clone, DeserializeKey, SilentDisplay, SilentDebug, SerializeKey)]
-pub struct X25519StaticPrivateKey(x25519_dalek::StaticSecret);
-
-/// An x25519 public key
-#[derive(Clone, Deref)]
-pub struct X25519PublicKey(x25519_dalek::PublicKey);
-
-/// An x25519 public key to match the X25519Static key type, which
-/// dereferences to an X25519PublicKey
-#[derive(Clone, DeserializeKey, Eq, Hash, PartialEq, SerializeKey)]
-pub struct X25519StaticPublicKey(X25519PublicKey);
-
-/// An x25519 shared key
-#[derive(SilentDisplay, SilentDebug)]
-pub struct X25519SharedKey(x25519_dalek::SharedSecret);
-
-/////////////////////////
-// X25519EphemeralPrivateKey Traits //
-/////////////////////////
-
-impl Uniform for X25519EphemeralPrivateKey {
-    fn generate<R>(rng: &mut R) -> Self
-    where
-        R: ::rand::SeedableRng + ::rand::RngCore + ::rand::CryptoRng,
-    {
-        X25519EphemeralPrivateKey(x25519_dalek::EphemeralSecret::new(rng))
-    }
-}
-
-impl PrivateKey for X25519EphemeralPrivateKey {
-    type PublicKeyMaterial = X25519PublicKey;
-}
-
-impl ExchangeKey for X25519EphemeralPrivateKey {
-    type DHPublicKeyMaterial = X25519PublicKey;
-    type DHSharedKeyMaterial = X25519SharedKey;
-
-    // Diffie-Hellman exchange
-    fn dh(self, their_public: &X25519PublicKey) -> X25519SharedKey {
-        let shared_secret = self.0.diffie_hellman(&their_public.0);
```

Comment:
I don't see a reason to delete this, especially as you are planning on writing our own noise.

Reply:
but we really should have a test that this actually works for a generic ExchangeKey

Reply:
why have this in our code if we already have this in dalek?

Reply:
Note that one of the reasons we had wrappers is to be able to perform extra validation on keys (if we decide in the future), i.e., low order or mixed order keys. This is ongoing discussion with dalek authors as well.

I.e., one of the properties that a malicious party can break is "key control", where none of the parties should be able to predict the shared key before ephemeral public keys are exchanged. It's not a big threat (if it's considered one), but we should document that if we allow it.

Reply:
It's a different perspective, as the rest of crypto schemes are self-contained, which would allow them to stand on their own as dependencies in other projects. This usually means 
wrapping of existing lib or tools (or implement from scratch) 
+
extra validation and control logic 
+ 
compatibility with other traits in the crypto crate (for consistency)
+ 
carrying all of the logic on how to use them (to sign, to encrypt, to DH).

This one lacks the last element, but I now realize it's fine as x25519 could be used for encryption, for DH, even for signatures. I'm ok with the fact that here we have a key wrapper, while the rest (ed25519, hash, hkdf, multi_ed25519) are full schemes. I hope we revisit this in the future and implement the execution logic as well, i.e., the noise DH/ECIES scheme should be under crypto crate if we would like crypto to stand on its own.